spring-grpc-spring-boot-autoconfigure/src/main/java/org/springframework/grpc/autoconfigure/server/GrpcServerAutoConfiguration.java

```diff
@@ -17,36 +17,51 @@
 
 import java.util.List;
 
-import org.springframework.boot.autoconfigure.AutoConfigureBefore;
-import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import io.grpc.BindableService;
+
+import org.springframework.beans.factory.ObjectProvider;
+import org.springframework.boot.autoconfigure.AutoConfiguration;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
 import org.springframework.context.annotation.Bean;
-import org.springframework.context.annotation.Configuration;
 import org.springframework.grpc.server.DefaultGrpcServerFactory;
 import org.springframework.grpc.server.GrpcServerConfigurer;
 import org.springframework.grpc.server.GrpcServerFactory;
+import org.springframework.grpc.server.lifecycle.GrpcServerLifecycle;
 
-import io.grpc.BindableService;
-import io.grpc.ServerServiceDefinition;
-
-@Configuration(proxyBeanMethods = false)
-@ConditionalOnMissingBean(GrpcServerFactory.class)
-@ConditionalOnBean(BindableService.class)
-@AutoConfigureBefore(GrpcServiceAutoConfiguration.class)
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for gRPC server-side components.
+ *
+ * @author David Syer
+ * @author Chris Bono
+ */
+@AutoConfiguration
+@ConditionalOnClass(BindableService.class)
 @EnableConfigurationProperties(GrpcServerProperties.class)
-public class GrpcServerFactoryAutoConfiguration {
+public class GrpcServerAutoConfiguration {
+
+	private final GrpcServerProperties properties;
+
+	GrpcServerAutoConfiguration(GrpcServerProperties properties) {
+		this.properties = properties;
+	}
 
 	@ConditionalOnMissingBean(GrpcServerFactory.class)
 	@Bean
-	public DefaultGrpcServerFactory<?> defaultGrpcServerFactory(final GrpcServerProperties properties,
-			final GrpcServiceDiscoverer serviceDiscoverer, final List<GrpcServerConfigurer> serverConfigurers) {
-		final DefaultGrpcServerFactory<?> factory = new DefaultGrpcServerFactory<>(properties.getAddress(),
-				properties.getPort(), serverConfigurers);
-		for (final ServerServiceDefinition service : serviceDiscoverer.findGrpcServices()) {
-			factory.addService(service);
-		}
+	DefaultGrpcServerFactory<?> defaultGrpcServerFactory(ObjectProvider<BindableService> grpcServicesProvider,
```

Comment:
I removed the `GrpcServiceDiscoverer` concept in favor of an `ObjectProvider<BindableService>`. We can introduce a discoverer later when needed is my thinking.

Reply:
I'm not 100% convinced about `ObjectProvider` (a simple `List` is easier to test), but this seems like a reasonable point generally.

Reply:
Most of the Spring Boot auto-configs are now using OP vs List and it is pretty easy to test w/ auto-configuration tests by setting up BS beans in the context and making sure those come into the server factory.